FastParallelEnvironment

emailweixu opened the pull request:
Design:

`FastParallelEnvironment` uses `_penv.ParallelEnvironment` (implemented in C++)    to coordinate step() and reset(). Each ``ProcessEnvironment`` maintains one ``_penv.ProcessEnvironmentCaller`` in the main process and one ``_penv.ProcessEnvironment`` in the client process.

In the client process, ``_penv.ProcessEnvironment.worker()`` runs in a loop to wait for jobs from either ``_penv.ParallelEnvironment`` or ``_penv.ProcessEnvironmentCaller``.

There are 4 types of job:

- step: step the environment. Sent from ``_penv.ParallelEnvironment``. The result is communicated back using shared memory.

- reset: reset the environment. Sent from ``_penv.ParallelEnvironment``. The result is communicated back using shared memory.

- close: close the environment. Sent from ``_penv.ProcessEnvironmentCaller``. This will cause the worker to finish and quit the process.

- call: access other methods of the environment. Sent from ``_penv.ProcessEnvironmentCaller``. This takes advantage of the pipe mechanism used by  the ``ParallelAlfEnvironment``. This is achieved by calling ``call_handler`` to do communication using python pipe. The reason of using the original pipe mechanism for other types of communication is that it is not easy to handle communication of unknow size using shared memory. 

Tested with gym breakout and hobot low level control env, the speed up is about 15%. Not as fast as what I expected.

emailweixu committed:
FastParallelEnvironment

breakds commented on alf/environments/process_environment.py:
```diff
@@ -45,7 +47,13 @@ class _MessageType(Enum):
+            env_constructor,
+            env_id=None,
+            flatten=False,
+            fast=False,
```
Need to update the docstring.

emailweixu replied:
udpated

breakds commented on alf/environments/process_environment.py:
```diff
@@ -110,8 +99,51 @@ def _worker(conn, env_constructor, env_id=None, flatten=False):
         conn.close()
 
 
+def process_call(conn, env, flatten, action_spec):
```
As far as I understand, `handle_process_call` sounds like a more accurate name?

breakds replied:
Sorry, I actually mistook `process` (as a verb) with `process` (as a noun, the program process). Still, `handle_call` would be better.

breakds commented on alf/environments/process_environment.py:
```diff
@@ -110,8 +99,51 @@ def _worker(conn, env_constructor, env_id=None, flatten=False):
 class ProcessEnvironment(object):
-    def __init__(self, env_constructor, env_id=None, flatten=False):
+    def __init__(self,
+                 env_constructor,
+                 env_id=None,
+                 flatten=False,
+                 fast=False,
+                 num_envs=0,
+                 name=""):
```
Update docstring.

emailweixu replied:
fixed.

breakds commented on alf/environments/parallel_environment.cpp:
```diff
@@ -0,0 +1,501 @@
+  bip::message_queue::remove((name_ + ".ready_queue").c_str());
+}
+
+class ProcessEnvironmentCaller {
```
I think the name `Caller` is a bit confusing - how about `ProcessEnvironmentProxy` or `ProcessEnvironmentClient` for better readability for the future reader?

emailweixu replied:
The reason for this name is that it only send "call" job to `_penv.ProcessEnvironment`

breakds commented on alf/environments/parallel_environment.cpp:
```diff
@@ -0,0 +1,501 @@
+  ret.reserve(list.size());
+  for (size_t i = 0; i < list.size(); ++i) {
+    ret.push_back(list[i]);
+  }
```
Why is it necessary to convert the `py::list` to `std::vector`?

emailweixu replied:
changed to py::list.

breakds commented on alf/environments/parallel_environment.cpp:
```diff
@@ -0,0 +1,501 @@
+    job_queues_.emplace_back(std::make_unique<bip::message_queue>(
+        bip::open_or_create,
+        (name + ".job_queue." + std::to_string(i)).c_str(),
+        3,
```
How is `3` picked for max number of messages?

emailweixu replied:
comment added.

breakds commented on alf/environments/parallel_environment.cpp:
```diff
@@ -0,0 +1,501 @@
+ protected:
+  int num_spare_envs_;
+  std::vector<int> original_env_ids_;  // [num_envs]
+  std::vector<uint8_t> reseted_;       // [num_envs]
```
Is there something wrong if we type this as `vector<bool>`?

emailweixu replied:
vector<bool> behaves differently from other vector (https://cplusplus.com/reference/vector/vector-bool/). It's more like bit array, which is slower.

breakds commented on alf/environments/parallel_environment.cpp:
```diff
@@ -0,0 +1,501 @@
+    int env_id;
+    ready_queue_.receive(&env_id, sizeof(env_id), recvd_size, priority);
+    assert(recvd_size == sizeof(env_id));
+    if (step_type_buf_[env_id] == (int32_t)StepType::last) {
```
Nit: `static_cast` over direct cast.

emailweixu replied:
updated.

breakds commented on alf/environments/parallel_environment.cpp:
```diff
@@ -0,0 +1,501 @@
+      env_.attr("close")();
+      break;
+    } else if (job.type == JobType::call) {
+      call_handler_();
```
I am still a bit confused by this. It seems that `call_handler_()` is just the original python `ProcessEnvironment`'s handler. Why do we still need it? This has a connection with a follow-up question below.

emailweixu replied:
Use call_handler to handle other types of communication with unknown size.

breakds commented on alf/environments/process_environment.py:
```diff
@@ -214,6 +252,8 @@ def __getattr__(self, name):
             Value of the attribute.
         """
         assert self._conn, "Run ProcessEnvironment.start() first"
+        if self._fast:
+            self._penv.call()
```
This is a part that is unclear to me. By design, if we delegate all the `main process` / `child process` communication to shared memory buffers in c++, why do we still need to use `Pipe` in fast mode? To put my question in a better way, isn't it possible to make the design such that Pipe is no longer needed in fast mode?

emailweixu replied:
Updated the description about "call" job in the PR description.

breakds replied:
Very clear now. Thank you!

breakds reviewed (commented):
Sorry it took quite some time to read through the code and try to find answers to my questions. Here are more comments/questions.

hnyu commented on setup.py:
```diff
@@ -15,6 +15,9 @@
 import os
 
+assert os.system(
+    "python3 alf/environments/make_penv.py") == 0, "Fail to build _penv"
```
Does this work with pypi package or pip install from ALF repo? A setup file like cnest might be needed https://github.com/HorizonRobotics/cnest/blob/master/setup.py . Feel like a separate penv python repo could be created.

emailweixu replied:
fixed

emailweixu committed:
Address comments

Also fixed a bug of not copying step and reset result.

hnyu commented on alf/environments/parallel_environment_test.py:
```diff
@@ -173,10 +178,10 @@ def test_non_blocking_reset_with_spare_envs(self):
-            sleep_time - 0.01,
-            msg=(f'Step without spare envs took {step4_time}, too short'))
+            sleep_time - 0.1,
+            msg=(f'Step witho spare envs took {step4_time}, too short'))
```
Why changing to assertLessEqual but the message is still 'short'?

emailweixu replied:
fixed

emailweixu committed:
Address comments

emailweixu committed:
Fix unittests

hnyu commented on alf/environments/parallel_environment.cpp:
```diff
@@ -44,31 +44,25 @@ namespace py = pybind11;
   return ((x + addr_alignment - 1) / addr_alignment) * addr_alignment;
 }
 
 // TODO(emailweixu): directly use cnest c++ function
```
remove TODO?

breakds replied:
I think Wei means using the one directly from c++, not via python binding.

emailweixu replied:
Yes, that's what I mean.

hnyu commented on alf/environments/parallel_environment.cpp:
```diff
@@ -44,31 +44,25 @@ namespace py = pybind11;
+  return py::list(flatten(nest));
 }
 
 // TODO(emailweixu): directly use cnest c++ function
```
remove TODO?

hnyu commented on setup.py:
```diff
@@ -57,6 +54,15 @@
+    cmdclass={'build_ext': build_ext},
     extras_require={
         'metadrive': ['metadrive-simulator==0.2.5.1', ],
     },
```
You need to include the compiled so file in the 'package_data'? Otherwise an installed alf package will miss the so file.

emailweixu replied:
it is handled automatically

emailweixu committed:
Merge branch 'pytorch' of https://github.com/HorizonRobotics/alf into PR_fast_parallel_env

emailweixu committed:
Improve setup.py

pybind11 is not always available. Add it as a prerequisite for setup in pyproject.toml
